File: LeetCode/Java/144-tree-二叉树的前序遍历.java
/*
 * @lc app=leetcode.cn id=144 lang=java
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start

class Solution {
    private List<Integer> res = new ArrayList();

    public List<Integer> preorderTraversal(TreeNode root) {
        _preorder(root);
        return res;
    }

    private void _preorder(TreeNode root) {
        if (root == null)
            return;
        res.add(root.val);
        _preorder(root.left);
        _preorder(root.right);
    }
}
// @lc code=end

File: LeetCode/C/144-tree-二叉树的前序遍历.cpp
/*
 * @lc app=leetcode.cn id=144 lang=cpp
 *
 * [144] 二叉树的前序遍历
 */
#include "main.h"
// @lc code=start
class Solution {
  public:
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> res;

        function<void(TreeNode *)> pre = [&](TreeNode *node) {
            if (node == nullptr)
                return;
            res.push_back(node->val);
            pre(node->left);
            pre(node->right);
        };

        pre(root);
        return res;
    }
};
// @lc code=end
